design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,753 @@
+* A `composite` object (required). This is the XR being reconciled.
+* A `resources` array (optional). Zero or more composed resources.
+* A `results` array (optional). Used to communicate information about the result
+  of a function, including warnings and errors.
+
+Each function:
+
+1. Takes its config, XR, and zero or more composed resources as input.
+1. Optionally mutates the XR's metadata, spec, and/or status.
+1. Optionally mutates the array of composed resources. This may involve adding
+   or deleting composed resources, as well as mutating the metadata and spec of
+   any entry in the array. Updates to the status of composed resources are not
+   allowed and will be silently discarded.
+
+This allows the output of one function to be the input to the next. The first
+function in the pipeline would be supplied with either:
+
+1. Only the XR. This would be the case for newly created XRs whose Composition
+   did not include a `resources` array.
+1. The XR and one or more resources. This would be the case if either:
+   1. The Composition included both a `resources` array and a `functions` array.
+      In this case the `resources` array is processed first, then passed to
+      `functions` for further processing.
+   1. This was not a newly created XR, but instead an XR that had already
+      created its composed resources and was now being updated.
+
+In the case of `Container` functions this input and output would correspond to
+stdin and stdout. This may differ for future function implementations; for
+example a webhook function implementation might take the input as an HTTP PUT
+request body and return the output as a response body. Crossplane would be
+responsible for reading stdout from the final function and applying its changes
+to the relevant XR and subsequent composed resources.
+
+An example function that responded by asking Crossplane to create (compose) a
+`CloudSQLInstance` would do so by returning the following `FunctionIO`:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: FunctionIO
+config:
+  apiVersion: database.example.org/v1alpha1
+  kind: Config
+  metadata:
+    name: cloudsql
+  spec:
+    version: POSTGRES_9_6
+composite:
+  apiVersion: database.example.org/v1alpha1
+  kind: XPostgreSQLInstance
+  metadata:
+    name: my-db
+  spec:
+    parameters:
+      storageGB: 20
+    compositionSelector:
+      matchLabels:
+        provider: gcp
+resources:
+- name: cloudsqlinstance
+  resource:
+    apiVersion: database.gcp.crossplane.io/v1beta1
+    kind: CloudSQLInstance
+    spec:
+      forProvider:
+        databaseVersion: POSTGRES_9_6
+        region: us-central1
+        settings:
+          tier: db-custom-1-3840
+          dataDiskType: PD_SSD
+          dataDiskSizeGb: 20
+      writeConnectionSecretToRef:
+        namespace: crossplane-system
+        name: cloudsqlpostgresql-conn
+  connectionDetails:
+  - name: hostname
+    fromConnectionSecretKey: hostname
```

Comment:
From @turkenh at https://gist.github.com/negz/c5a5b4f82b06882906b08af377a609a4?permalink_comment_id=4329791#gistcomment-4329791:

> Actually I am wondering if we need to pass this connectionDetails field at all? If we want to pass existing connection secrets as input and expect XR connection secret as output (as we discussed in a separate thread), I believe outputting the final XR secret should be enough.

I was thinking the same originally, but I don't think we can rely on Functions outputting the XR secret. Consider that:

* Many (most?) Functions will simply be logic that takes an XR as input and produces (the manifests of) one or more MRs as output. They won't actually create those MRs, so they won't know what their connection details are.
* Sometimes a contemporary P&T style `resources` array will be computed before the array of `functions`. How would the functions know to include any connection detail (mappings) from the `resources` array?

I think we need Functions to be able to handle at least what P&T style Composition can do, with regard to connection details. So that would be:

* After specifying a new MR that Crossplane should create, tell Crossplane how to derive XR connection details from that MR's connection details.
* Support returning an explicit XR connection detail (i.e. don't derive it, just use this value). This could use the same API as P&T `FromValue` style connection details, but the advantage of doing so in a Function is that the value could be somehow computed (e.g. by calling an external system).

Reply:
An alternative to including `connectionDetails` and `readinessChecks` would be to provide functions with enough context to figure this out themselves - i.e. providing functions with the status and connection details of composed resources.

Let's say for example we returned to an API more like `ResourceList` (so a big array of resources, with some way to differentiate between the XR, composed resources, and other types). We could:

First reconcile of a new XR:

1. Crossplane passes the XR as the only element in the `ResourceList` to a Function.
2. Function responds by adding one or more composed resources to the `ResourceList`, returns it to Crossplane.
3. Crossplane creates those composed resources.

Second reconcile of the above XR:

1. Crossplane passes the XR, and composed resources as elements in a `ResourceList` to a Function. It also passes the XR and composed resources' connection details - let's say as `kind: Secret`.
2. Function responds by mapping connection details from composed resource Secrets to XR Secret as appropriate.
3. Function updates XR status (probably derived from composed resource status) as appropriate.

I don't hate this alternative. I think the key upside is that it's much more flexible than the current `FunctionIO` approach, but it does have a few downsides (or at least quirks):

* Connection details aren't actually a `kind: Secret` if you use an external secret store (e.g. Vault). Representing them in functions as `kind: Secret` could be misleading or confusing to Function authors.
* Always having to include logic in your Function to derive XR connection details and status from composed resources could become repetitive and painful fast.
* It's more work for function authors to 'find things' in the array. For example you can't just look at the `composite` field - you need to iterate over all resources and find the one that's the XR.
* We'd need to create associations/relationships within the array - e.g. "this Secret represents the connection details for the composed resource named `foo`".
* We'd have to rely on annotations to represent types (e.g. XR vs composed vs connection details), names, and relationships

Reply:
I am on the same page regarding that `FunctionIO` approach works better than `ResourceList` approach.

If we need a more flexible solution while eliminating the quirks you've mentioned except the second one (unless it introduces others), a combined solution could be passing connection secret data under a `connectionData` field and expecting XR secret from `composite.connectionData`, instead of passing **how** to build it with `connectionDetails`.

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: FunctionIO
config:
  apiVersion: database.example.org/v1alpha1
  kind: Config
  metadata:
    name: cloudsql
  spec:
    version: POSTGRES_9_6
composite:
  resource:
    apiVersion: database.example.org/v1alpha1
    kind: XPostgreSQLInstance
    metadata:
      name: my-db
    spec:
      parameters:
        storageGB: 20
      compositionSelector:
        matchLabels:
          provider: gcp
  connectionData:
    dbURI: mysql://foo:bar@hostname:3306
resources:
- name: cloudsqlinstance
  resource:
    apiVersion: database.gcp.crossplane.io/v1beta1
    kind: CloudSQLInstance
    spec:
      forProvider:
        databaseVersion: POSTGRES_9_6
        region: us-central1
        settings:
          tier: db-custom-1-3840
          dataDiskType: PD_SSD
          dataDiskSizeGb: 20
      writeConnectionSecretToRef:
        namespace: crossplane-system
        name: cloudsqlpostgresql-conn
  connectionData:
    username: foo
    password: bar
    host: hostname
    port: 3306
  readinessChecks:
  - type: None
```

> Always having to include logic in your Function to derive XR connection details and status from composed resources could become repetitive and painful fast.

Not sure if it would be painful for connection secrets with an approach like above. But I can kind of get that it could be repetitive to build the readiness condition calculation in each function by checking all composed resources statuses instead of having a common logic. But I believe this might be acceptable as a trade off for more flexibility.

Reply:
@turkenh Maybe there's a best of both worlds approach? Something like:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: FunctionIO
config:
  apiVersion: database.example.org/v1alpha1
  kind: Config
  metadata:
    name: cloudsql
  spec:
    version: POSTGRES_9_6
composite:
  resource:
    apiVersion: database.example.org/v1alpha1
    kind: XPostgreSQLInstance
    metadata:
      name: my-db
    spec:
      parameters:
        storageGB: 20
      compositionSelector:
        matchLabels:
          provider: gcp
  # XR connection details can be specified by an entry in the resources array
  # too. Allowing them to be specified here decouples them from a specific
  # resources array entry - e.g. it allows a function to 'build' an connection
  # detail for an XR by merging multiple resource connection details, or from
  # some entirely unrelated source.
  connectionDetails:
  - type: FromValue
    name: uri
    value: postgresql://admin:password@db.example.org:5432
  # Connection details that already exist - i.e. have been persisted to and read
  # back from a secret store. Read-only to the function.
  existingConnectionDetails:
  - name: uri
    value: postgresql://admin:password@db.example.org:5432
resources:
- name: cloudsqlinstance
  resource:
    apiVersion: database.gcp.crossplane.io/v1beta1
    kind: CloudSQLInstance
    spec:
      forProvider:
        databaseVersion: POSTGRES_9_6
        region: us-central1
        settings:
          tier: db-custom-1-3840
          dataDiskType: PD_SSD
          dataDiskSizeGb: 20
      writeConnectionSecretToRef:
        namespace: crossplane-system
        name: cloudsqlpostgresql-conn
  connectionDetails:
  - type: FromConnectionSecretKey
    name: hostname
    fromConnectionSecretKey: hostname
  - type: FromValue
    name: username
    value: admin
  # Connection details that already exist - i.e. have been persisted to and read
  # back from a secret store. Read-only to the function.
  existingConnectionDetails:
  - name: hostname
    value: db.example.org
  - name: username
    value: admin
  readinessChecks:
  - type: None
```

Under this approach functions can set `connectionDetails` to configure how XR connection details are set (either `FromConnectionSecretKey` or `FromValue`), but they also get to see any existing (already persisted) connection details - which they could use to create new `FromValue` entries in the `connectionDetails` array.

Reply:
@turkenh The main thing I don't like about my most recent suggestion above is that it starts to feel like it's mixing desired state and actual state without clear differentiation.

One way to address that would be to nest everything in `FunctionIO` under spec and status, like a typical Kubernetes object. The `status` object would be read-only to functions, and exist strictly to provide them with context about the actual state of the world (as opposed to the desired state, as specified by the Composition's `resources` array, and/or the logic of any previous functions that have run before them.

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: FunctionIO
spec:
  config:
    apiVersion: database.example.org/v1alpha1
    kind: Config
    metadata:
      name: cloudsql
    spec:
      version: POSTGRES_9_6
  composite:
    resource:
      apiVersion: database.example.org/v1alpha1
      kind: XPostgreSQLInstance
      metadata:
        name: my-db
      spec:
        parameters:
          storageGB: 20
        compositionSelector:
          matchLabels:
            provider: gcp
    connectionDetails:
    - type: FromValue
      name: uri
      value: postgresql://admin:password@db.example.org:5432
  resources:
  - name: cloudsqlinstance
    resource:
      apiVersion: database.gcp.crossplane.io/v1beta1
      kind: CloudSQLInstance
      spec:
        forProvider:
          databaseVersion: POSTGRES_9_6
          region: us-central1
          settings:
            tier: db-custom-1-3840
            dataDiskType: PD_SSD
            dataDiskSizeGb: 20
        writeConnectionSecretToRef:
          namespace: crossplane-system
          name: cloudsqlpostgresql-conn
    connectionDetails:
    - name: hostname
      fromConnectionSecretKey: hostname
    readinessChecks:
    - type: None
status:
  composite:
    resource:
      apiVersion: database.example.org/v1alpha1
      kind: XPostgreSQLInstance
      metadata:
        name: my-db
      spec:
        parameters:
          storageGB: 20
        compositionSelector:
          matchLabels:
            provider: gcp
      status:
        state: ACTIVE
        conditions:
        - type: Ready
          status: True
    connectionDetails:
    - name: uri
      value: postgresql://admin:password@db.example.org:5432
  resources:
  - name: cloudsqlinstance
    resource:
      apiVersion: database.gcp.crossplane.io/v1beta1
      kind: CloudSQLInstance
      spec:
        forProvider:
          databaseVersion: POSTGRES_9_6
          region: us-central1
          settings:
            tier: db-custom-1-3840
            dataDiskType: PD_SSD
            dataDiskSizeGb: 20
        writeConnectionSecretToRef:
          namespace: crossplane-system
          name: cloudsqlpostgresql-conn
      status:
        atProvider:
          state: ACTIVE
        conditions:
        - type: Ready
          status: True
    connectionDetails:
    - name: hostname
      value: db.example.org
```

Reply:
I think we are getting there 🙂

Looking better than what I suggested above, but I am a bit confused who reads/writes specs vs status, i.e. what is input to the function and what is output of the function. I am wondering if we could make it more clear with the following:

Input to the function:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: FunctionIO
config:
  apiVersion: database.example.org/v1alpha1
  kind: Config
  metadata:
    name: cloudsql
  spec:
    version: POSTGRES_9_6
observed:
  composite:
    resource:
      apiVersion: database.example.org/v1alpha1
      kind: XPostgreSQLInstance
      metadata:
        name: my-db
      spec:
        parameters:
          storageGB: 20
        compositionSelector:
          matchLabels:
            provider: gcp
      status:
        state: ACTIVE
        conditions:
        - type: Ready
          status: True
    connectionDetailsData:
    - name: uri
      value: postgresql://admin:password@db.example.org:5432
  resources:
  - name: cloudsqlinstance
    resource:
      apiVersion: database.gcp.crossplane.io/v1beta1
      kind: CloudSQLInstance
      spec:
        forProvider:
          databaseVersion: POSTGRES_9_6
          region: us-central1
          settings:
            tier: db-custom-1-3840
            dataDiskType: PD_SSD
            dataDiskSizeGb: 20
        writeConnectionSecretToRef:
          namespace: crossplane-system
          name: cloudsqlpostgresql-conn
      status:
        atProvider:
          state: ACTIVE
        conditions:
        - type: Ready
          status: True
    connectionDetailsData:
    - name: hostname
      value: db.example.org
```

Output of the function:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: FunctionIO
desired:
  composite:
    resource:
      apiVersion: database.example.org/v1alpha1
      kind: XPostgreSQLInstance
      metadata:
        name: my-db
      spec:
        parameters:
          storageGB: 20
        compositionSelector:
          matchLabels:
            provider: gcp
    connectionDetails:
    - type: FromValue
      name: uri
      value: postgresql://admin:password@db.example.org:5432
  resources:
  - name: cloudsqlinstance
    resource:
      apiVersion: database.gcp.crossplane.io/v1beta1
      kind: CloudSQLInstance
      spec:
        forProvider:
          databaseVersion: POSTGRES_9_6
          region: us-central1
          settings:
            tier: db-custom-1-3840
            dataDiskType: PD_SSD
            dataDiskSizeGb: 20
        writeConnectionSecretToRef:
          namespace: crossplane-system
          name: cloudsqlpostgresql-conn
    connectionDetails:
    - name: hostname
      fromConnectionSecretKey: hostname
    readinessChecks:
    - type: None
```

What I did was:
- Moved `spec.config` up to top level as `config`.
- Renamed `spec` as `desired`.
- Renamed `status` as `observed`.
- Renamed `status.connectionDetails` as `observed.ConnectionDetailsData` (not mandatory but to indicate they are different types).

Reply:
Also with the above, `status` could still be used for error reporting by the function. `input`/`output` could be another alternative to `observed`/`desired`.

Reply:
@turkenh I considered `observed` and `desired` too - I actually like them more than `spec` and `status` but I'm not sure they're "better" enough to warrant diverging from the typical `spec` and `status` names.

> I am a bit confused who reads/writes specs vs status, i.e. what is input to the function and what is output of the function. 

FWIW I don't think `input` and `output` make sense in the context of a pipeline of functions, where the output of one function is the input to another. Instead I think of all functions in the pipeline as working together to "accumulate" a desired state, by each mutating the spec.

So, under my proposal:

* All functions may write to `spec` - the "desired state" as accumulated by the functions.
* No functions may write to `status` - the "observed state" used to inform the functions.

I do like the idea of moving `config` to the top level, since it's different from the function's desired state. As I write this out I'm thinking I do like `observed` and `desired` better than `spec` and `status` because they have different meaning from the typical `spec` and `status` - it's probably confusing that `status` isn't actually the status of a function.

I'll push a commit to reflect the changes we've discussed here.

Reply:
> I'll push a commit to reflect the changes we've discussed here.

@turkenh Done. Let me know what you think (feel free to resolve this thread if it looks good.